enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -3976,6 +3979,152 @@ class AdjointGenerator
+                    uncacheable_args.find(xfuncarg)->second;
+      bool ycache = !gutils->isConstantValue(call.getArgOperand(1)) &&
+                    uncacheable_args.find(yfuncarg)->second;
+      if ((Mode == DerivativeMode::ReverseModeCombined ||
+           Mode == DerivativeMode::ReverseModePrimal) &&
+          (xcache || ycache)) {
+        BuilderZ.SetInsertPoint(gutils->getNewFromOriginal(&call));
+        Value *arg1, *arg2;
+        auto size = ConstantExpr::getSizeOf(innerType);
+        if (xcache) {
+          auto dmemcpy = getOrInsertMemcpyStrided(
+              *BuilderZ.GetInsertBlock()->getParent()->getParent(),
+              PointerType::getUnqual(innerType), 0, 0, call.getArgOperand(2));
+          auto malins = CallInst::CreateMalloc(
+              gutils->getNewFromOriginal(&call), size->getType(), innerType,
+              size, call.getArgOperand(0), nullptr, "");
+          arg1 =
+              BuilderZ.CreateBitCast(malins, call.getArgOperand(1)->getType());
+          SmallVector<Value *, 3> args;
+          args.push_back(arg1);
+          args.push_back(gutils->getNewFromOriginal(call.getArgOperand(1)));
+          args.push_back(call.getArgOperand(0));
+          BuilderZ.CreateCall(dmemcpy, args);
+        }
+        if (ycache) {
+          auto dmemcpy = getOrInsertMemcpyStrided(
+              *BuilderZ.GetInsertBlock()->getParent()->getParent(),
+              PointerType::getUnqual(innerType), 0, 0, call.getArgOperand(4));
+          auto malins = CallInst::CreateMalloc(
+              gutils->getNewFromOriginal(&call), size->getType(), innerType,
+              size, call.getArgOperand(0), nullptr, "");
+          arg2 =
+              BuilderZ.CreateBitCast(malins, call.getArgOperand(3)->getType());
+          SmallVector<Value *, 3> args;
+          args.push_back(arg2);
+          args.push_back(gutils->getNewFromOriginal(call.getArgOperand(3)));
+          args.push_back(call.getArgOperand(0));
+          BuilderZ.CreateCall(dmemcpy, args);
+        }
+        if (xcache && ycache) {
+          auto valins1 = BuilderZ.CreateInsertValue(undefinit, arg1, 0);
+          cacheval = BuilderZ.CreateInsertValue(valins1, arg2, 1);
+        } else if (xcache)
+          cacheval = arg1;
+        else if (ycache)
+          cacheval = arg2;
+        gutils->cacheForReverse(BuilderZ, cacheval,
+                                getIndex(&call, CacheType::Tape));
+      }
+      if (Mode == DerivativeMode::ReverseModeCombined ||
+          Mode == DerivativeMode::ReverseModeGradient) {
+        IRBuilder<> Builder2(call.getParent());
+        getReverseBuilder(Builder2);
+        auto derivcall = gutils->oldFunc->getParent()->getOrInsertFunction(
+            dfuncName, Builder2.getVoidTy(), Builder2.getInt32Ty(), innerType,
+            call.getArgOperand(1)->getType(), Builder2.getInt32Ty(),
+            call.getArgOperand(3)->getType(), Builder2.getInt32Ty());
+        Value *structarg1 = lookup(
```

Comment:
For performance, it's good to avoid a lookup if not necessary. Can you initialize this to nullptr, then if it's not set from the cache, do this lookup?

Reply:
Bumping this. Also its wise to avoid a lookup if it's not used to avoid unnecessary caching.